Automattic/wordpress-activitypub
# Improve transformer

Adds "improved" error handling to code that uses the `Transformer` class(es). (Really, all it does is add a bunch of `is_wp_error( $transformer )` checks!)

(The current version of the plugin has some checks in place, but not enough to prevent fatal errors down the line, _should one ever attempt to pass an invalid object or transformer_ [to the rest of the code].)

As a bonus, it allows _somewhat experienced PHP developers_ to _exclude_ certain posts (and, potentially, comments) from federation. (Which is an oft-requested feature!)

## Proposed changes:
* Add better "transformer error" handling to JSON templates and ActivityPub dispatcher
* Add better "transformer error" handling to outbox and collection endpoint
    * As a result, "method chaining" is no longer possible (I think); I'd suggest using the "Factory" everywhere, make things a bit simpler
        * If this is somehow not an option, then `Activitypub\Transformer\Base::transform()` has to become filterable, too (to prevent "unwanted" posts from showing in outboxes)
            * This PR currently includes these changes, but they could be scrapped if this function is no longer used
        * If it is, maybe deprecate this function in favor of `Factory::get_transformer()`? 😬

### Other information:
* As mentioned above, this _could_, potentially, be used to _exclude posts from federation_.
* While I see some challenges (related to [#668](https://github.com/Automattic/wordpress-activitypub/issues/668)) using the filter _for that purpose_, that's ... a different issue, really. (Probably should not make things this confusing ... but I wanted to mention it regardless.)
* The only "real" downside I see is that by _allowing outboxes and collections to be filterable_, that the per-page counts will no longer "be correct" (But only _if such a filter is actually in use_. Also, I'm not sure if any software currently relies on outboxes, or if they expect a page to hold a certain number of items _always_.)
    * Reason for this is filtering happens _after_ fetching the items from the database ... (There's currently no way to filter "outbox _queries_." Unless maybe we use the `federated` custom field for that.)
    * So that _could_ be a reason to leave (only) the outbox/collection endpoints as they currently are.
 
- [ ] I did not run or add automated tests (no Docker on this machine, although I don't expect problems, the code does not change the plugin's default behavior) 😬
- [x] I absolutely tested in production

## Testing instructions:
I'm going to just outline what I did (and do):
- Add an mu-plugin like this one:
```
<?php

add_filter( 'activitypub_transformer', function ( $transformer, $obj ) {

	if ( $obj instanceof \WP_Post && in_category( 'disable-activitypub', $obj ) ) {
		return new \WP_Error( 'unsupported_object', __( 'This object does not support ActivityPub.', 'activitypub' ) );
	}

	return $transformer;

}, 10, 2 );
```
- Create a new post, and make sure it has the `disable-activitypub` category
- Save it as draft, to _ensure Gutenberg has applied this category_ before "federation" is scheduled
- Publish the post
- Watch _no_ outgoing POST requests happen
- Finally, try searching for the post's URL in, e.g., Mastodon. Nothing should show up.
- Try visiting the `?activitypub` version of the posts, in your own browser. You should "simply" see the post, _not_ JSON.
- If you used such a mu-plugin (i.e., an `activitypub_transformer` filter that does _not_ return an actual transformer) with the currently released version of the plugin, you'd instead get a fatal PHP error.

The mu-plugin can "easily" be adapted to "exclude" post formats, posts with certain custom fields, and so on. (The only requirement is that these are stored before scheduling runs, which _could be tricky when you use the block editor_. As per the linked issue above.)
